SagaLabs-Manager-Client/src/main/java/sagalabsmanagerclient/VMSnapshot.java

```diff
@@ -0,0 +1,214 @@
+
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class VMSnapshot extends AzureUtils {
+    // Get the gallery resource ID from the Azure Key Vault secret
+    static String vmImageGalleryResourceGroup = "SL-vmImages";
+    static String vmImageGalleryNameDefault = "SagalabsVM";
+
+    public static void takeSnapshots(ArrayList<MachinesVM> vms, String version, String galleryName, Runnable onStartCallback) {
+        // Get the Azure resource manager instance
+        AzureResourceManager azure = AzureLogin.getAzure();
+
+        // Create a thread pool with a fixed number of threads
+        int numThreads = Math.min(10, vms.size()); // Limit the number of threads to a reasonable value
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+
+        // Loop over all VMs in the list
+        List<Future<Void>> futures = new ArrayList<>();
+        AtomicInteger machinesStarted = new AtomicInteger(0);
+        int totalMachines = vms.size();
+
+        for (MachinesVM vm : vms) {
+            // Submit a task for each VM to be processed concurrently
+            Future<Void> future = executorService.submit(() -> {
+                VirtualMachine azureVm = azure.virtualMachines().getById(vm.getAzureId());
+
+                // Retrieve the VM instance view
+                if (azureVm != null) {
+                    azureVm.refreshInstanceView();
+                    List<InstanceViewStatus> statuses = azureVm.instanceView().statuses();
+
+                        takeSpecializedSnapshot(vm.getResourceGroup(), vm.getVmName(), version, vmImageGalleryResourceGroup, galleryName, totalMachines, machinesStarted, onStartCallback);
+
+                }
+
+
+
+                return null;
+            });
+            futures.add(future);
+        }
+
+        // Wait for all tasks to complete
+        boolean onStartCallbackTriggered = false;
+        for (Future<Void> future : futures) {
+            try {
+                future.get();
+                if (!onStartCallbackTriggered && machinesStarted.get() == totalMachines) {
+                    onStartCallback.run();
+                    onStartCallbackTriggered = true;
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                System.out.println("Error processing VM: " + e.getMessage());
+            }
+        }
+
+        // Shutdown the executor service
+        executorService.shutdown();
+    }
+
+
+
+
+
+    private static String incrementVersion(String version) {
+        Pattern pattern = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d+)");
```

Comment:
Pls comment, hvad er det pattern?

Reply:
Pattern er digit,digit,digit, fordi azure har et versionsnummer til hvert image som hedder major.minor.patch. så man skal indtaste 1.0.1 eller 1.0.0. men faktisk bliver increment version ikke brugt lige nu, da brugeren i prencippet godt kan opdaterer den samme version "in place", uden at lave en ny version. det kan spare lidt plads i azure, men så har man ikke en historik af versionerne. Jeg lader den blive i koden indtil vi har lavet en god strategi for images i SagaLabs teamet